### PR TITLE
Added connections module and fixed a couple of bugs

### LINKED ISSUE #21: Remove uniqueness of the values in the Secret Store
Currently one can't create two keys with the same value because the column in the storage DB is marked unique.

### LINKED ISSUE #22: Bug in the usage example for the Secrets in the doc
The actual interface for the Secrets is
`Secrets(Path, str)`
not
`Secrets(str, str)`

### LINKED ISSUE #23: Add functions for creating connections to DB and BucketFS
Create functions that given the configuration as a Secrets object will create the following types of connections.

- pyexasol
- Exasol SQLAlchemy
- BucketFS

### LINKED ISSUE #26: Allow providing a default value in the `get` method of the Secret Store.
Make it in line with the similar dict interface, e.g. `get(key: str, value: Optional[str] = None)`

### PR DESCRIPTION
**Pull request by ahsimb**
closes #21 
closes #22 
closes #23 
closes #26

### CODE REVIEW

File: test/unit/test_connections.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+
+@unittest.mock.patch("pyexasol.connect")
+def test_open_pyexasol_connection(mock_connect, conf):
+    conf.save("SCHEMA", "IDA")
+
+    open_pyexasol_connection(conf)
+    mock_connect.assert_called_once_with(
+        dsn="24.134.96.2:8888", user="me", password="let_me_in", schema="IDA"
+    )
```
Comment (ckunki) [inline test/unit/test_connections.py:51]:
```suggestion
    dsn = f"{conf.EXTERNAL_HOST_NAME}:{conf.DB_PORT}"
    mock_connect.assert_called_once_with(
        dsn=dsn, user=conf.USER, password=conf.PASSWORD, schema="IDA"
    )
```

File: test/unit/test_connections.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+@unittest.mock.patch("pyexasol.connect")
+def test_open_pyexasol_connection_kwargs(mock_connect, conf):
+    open_pyexasol_connection(conf, connection_timeout=3, query_timeout=10)
+    mock_connect.assert_called_once_with(
+        dsn="24.134.96.2:8888",
+        user="me",
+        password="let_me_in",
+        connection_timeout=3,
+        query_timeout=10,
+    )
```
Comment (ckunki) [inline test/unit/test_connections.py:63]:
```suggestion
    dsn = f"{conf.EXTERNAL_HOST_NAME}:{conf.DB_PORT}"
    mock_connect.assert_called_once_with(
        dsn=dsn,
        user=conf.USER,
        password=conf.PASSWORD,
        connection_timeout=3,
        query_timeout=10,
    )
```

File: test/unit/test_connections.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        conf.save("CERTIFICATE_VALIDATION", "Yes")
+        conf.save("TRUSTED_CA", tmp_files[0].name)
+        conf.save("CLIENT_CERTIFICATE", tmp_files[1].name)
+        conf.save("PRIVATE_KEY", tmp_files[2].name)
+
+        open_pyexasol_connection(conf)
+        mock_connect.assert_called_once_with(
+            dsn="24.134.96.2:8888",
+            user="me",
+            password="let_me_in",
```
Comment (ckunki) [inline test/unit/test_connections.py:82]:
```suggestion
        dsn = f"{conf.EXTERNAL_HOST_NAME}:{conf.DB_PORT}"
        mock_connect.assert_called_once_with(
            dsn=dsn,
            user=conf.USER,
            password=conf.PASSWORD,
```

File: test/unit/test_connections.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        open_pyexasol_connection(conf)
+
+
+@unittest.mock.patch("sqlalchemy.create_engine")
+def test_open_sqlalchemy_connection(mock_create_engine, conf):
+    conf.save("SCHEMA", "IDA")
+
+    open_sqlalchemy_connection(conf)
+    mock_create_engine.assert_called_once_with(
+        "exa+websocket://me:let_me_in@24.134.96.2:8888/IDA"
```
Comment (ckunki) [inline test/unit/test_connections.py:109]:
```suggestion
    dsn = f"{conf.EXTERNAL_HOST_NAME}:{conf.DB_PORT}"
    mock_create_engine.assert_called_once_with(
        f"exa+websocket://{conf.USER}:{conf.PASSWORD}@{dsn}/IDA"        
```

File: test/unit/test_connections.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+@unittest.mock.patch("sqlalchemy.create_engine")
+def test_open_sqlalchemy_connection_ssl(mock_create_engine, conf):
+    conf.save("ENCRYPTION", "True")
+    conf.save("CERTIFICATE_VALIDATION", "False")
+
+    open_sqlalchemy_connection(conf)
+    mock_create_engine.assert_called_once_with(
+        "exa+websocket://me:let_me_in@24.134.96.2:8888"
+        "?ENCRYPTION=Yes&SSLCertificate=SSL_VERIFY_NONE"
```
Comment (ckunki) [inline test/unit/test_connections.py:121]:
```suggestion
    dsn = f"{conf.EXTERNAL_HOST_NAME}:{conf.DB_PORT}"
    mock_create_engine.assert_called_once_with(
        f"exa+websocket://{conf.USER}:{conf.PASSWORD}@{dsn}/IDA"  
        "?ENCRYPTION=Yes&SSLCertificate=SSL_VERIFY_NONE"
```

File: test/unit/test_connections.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    )
+
+
+@unittest.mock.patch("exasol.bucketfs.Service")
+def test_open_bucketfs_connection(mock_bfs_service, conf):
+    open_bucketfs_connection(conf)
+    mock_bfs_service.assert_called_once_with(
+        "http://24.134.96.2:6666",
```
Comment (ckunki) [inline test/unit/test_connections.py:129]:
```suggestion
    mock_bfs_service.assert_called_once_with(
        f"http://{conf.EXTERNAL_HOST_NAME}:{conf.BUCKETFS_PORT}",
```

File: test/unit/test_connections.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    def mock_get(self, key: str, default_value: Optional[str] = None) -> Optional[str]:
+        return self._params.get(key, default_value)
+
+    mock_conf = create_autospec(Secrets)
+    mock_conf._params = {}
+    mock_conf.save = types.MethodType(mock_save, mock_conf)
+    mock_conf.get = types.MethodType(mock_get, mock_conf)
+    mock_conf.EXTERNAL_HOST_NAME = "24.134.96.2"
+    mock_conf.HOST_PORT = "8888"
```
Comment (ckunki) [inline test/unit/test_connections.py:33]:
```suggestion
    mock_conf.DB_PORT = "8888"
```

File: test/unit/test_secret_store.py
Comment (ckunki) [inline test/unit/test_secret_store.py:1]: Please add a test verifying non-unique values.

File: test/unit/test_connections.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                "keyfile": tmp_files[2].name,
+            },
+        )
+
+
+@unittest.mock.patch("pyexasol.connect")
+def test_open_pyexasol_connection_error(mock_connect, conf):
+    conf.save("ENCRYPTION", "True")
+    conf.save("CERTIFICATE_VALIDATION", "Yes")
+    conf.save("TRUSTED_CA", "#%&")
```
Comment (ckunki) [inline test/unit/test_connections.py:97]:
```suggestion
    conf.save("TRUSTED_CA", "invalid or unknown CA")
```

**Review by ckunki — CHANGES_REQUESTED** [review]

File: test/unit/test_connections.py (hunk 8 above)
Comment (ahsimb) [reply test/unit/test_connections.py:100]: I like my filename more. A file with such a name is guaranteed not to exist.

File: test/unit/test_secret_store.py [hunk 9]
```diff
@@ -23,6 +23,14 @@ def test_value(secrets):
     assert secrets.get("key") == value
 
 
+def test_dup_values(secrets):
+    # Here we test that it is possible to save the same value with different keys
```
Comment (ckunki) [inline test/unit/test_secret_store.py:27]: :+1:

Comment (ckunki) [inline test/unit/test_secret_store.py:27]: :+1:

**Review by ckunki — APPROVED** [review]

**Pull request merged**